lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -5,10 +5,189 @@ define((require, exports, module) ->
+        acceptToken = true
+        if token.type == "storage.type" and token.value == "\\label"
+          curLine = session.getLine(tokenPosition.row)
+          bracketPosition = tokenPosition.column + "\\label".length
+          if curLine[bracketPosition] == "{"
+            argumentRange = ContextHandler.getMacrosArgumentRange(session, {row: tokenPosition.row, column: bracketPosition + 1})
+            if argumentRange?
+              acceptToken = false
+              labelParameters.push(session.getTextRange(argumentRange))
+              tokenIterator.stepTo(argumentRange.end.row, argumentRange.end.column + 1)
+
+        if acceptToken
+          tokenValues.push(token.value)
+
+        tokenIterator.stepForward()
+        token = tokenIterator.getCurrentToken()
+        tokenPosition = tokenIterator.getCurrentTokenPosition()
+
+      return [labelParameters, tokenValues.join("")]
+
+    constructor: (@editor, @popoverHandler, @equationRangeHandler, @getFormulaElement) ->
+      @jqEditorContainer = $(@editor.container)
+      @contextPreviewExists = false
+
+    getPopoverPosition: (row) -> {
+        top: "#{@editor.renderer.textToScreenCoordinates(row + 2, 1).pageY}px"
+        left: "#{@jqEditorContainer.position().left}px"
+      }
+
+    getCurrentFormula: ->
+      try
+        {row: startRow, column: startColumn} = @curRange.start
+        tokenIterator = new ConstrainedTokenIterator(@editor.getSession(), @curRange, startRow, startColumn)
+        tokenIterator.stepForward()
+        [labelParameters, equationString] = ContextHandler.getWholeEquation(@editor.getSession(), tokenIterator)
+        title = if labelParameters.length == 0 then "Formula" else labelParameters.join(", ")
+        return [title, katex.renderToString(equationString, ContextHandler.KATEX_OPTIONS)]
+      catch e
+        return ["Error!", e]
+
+    initPopover: => setTimeout((=>
+      popoverPosition = @getPopoverPosition(@getEquationEnd())
+      [title, rendered] = @getCurrentFormula()
+      @popoverHandler.show(@getFormulaElement(), title, rendered, popoverPosition)
+    ), 0)
+
+    getEquationEnd: ->
+      i = @editor.getCursorPosition().row
+      while LatexParsingContext.getContext(@editor.getSession(), i) == "equation"
+        i += 1
+      return i
+
+    updatePosition: =>
+      @popoverHandler.setPosition(@getFormulaElement(), @getPopoverPosition(@getEquationEnd()))
+
+    updateRange: ->
+      {row: cursorRow, column: cursorColumn} = @editor.getCursorPosition()
+      @curRange = @equationRangeHandler.getEquationRange(cursorRow, cursorColumn)
+
+    updatePopover: ->
+      if @contextPreviewExists
+        [title, rendered] = @getCurrentFormula()
+        @popoverHandler.setContent(@getFormulaElement(), title, rendered)
+
+    updateCallback: =>
+      if @lastChangeTime?
+        curTime = Date.now()
+        @currentDelayedUpdateId = setTimeout(@updateCallback, ContextHandler.UPDATE_DELAY - (Date.now() - @lastChangeTime))
```

Comment:
it does happen because the only guarantee which setTimeout provides is that code will be executed no early than the specified delay. It may execute later.

https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setTimeout#Reasons_for_delays_longer_than_specified